staging/kos/pkg/controllers/connectionagent/connection_agent.go

```diff
@@ -0,0 +1,1114 @@
+	// by VNI so their VNI can be used as the ID, but the local NetworkAttachments
+	// informer is not associated to a single VNI. Pick 0 because it's not a
+	// valid VNI value: there's no overlapping with the other IDs.
+	localAttsListerID = 0
+
+	// Name of the indexer used to match pre-existing network interfaces to
+	// network attachments.
+	ifcOwnerDataIndexerName = "ifcOwnerData"
+
+	// Field names of NetworkAttachments used to build field selectors.
+	attNodeField   = "spec.node"
+	attIPv4Field   = "status.ipv4"
+	attHostIPField = "status.hostIP"
+	attVNIField    = "status.addressVNI"
+
+	// resync period for Informers caches. Set
+	// to 0 because we don't want resyncs.
+	resyncPeriod = 0
+
+	// netFabricRetryPeriod is the time we wait before retrying when a
+	// network fabric operation fails while handling pre-existing interfaces.
+	netFabricRetryPeriod = time.Second
+
+	// The HTTP port at which Prometheus metrics are served.
+	// Pick an unusual one because the host's network namespace is used.
+	// See https://github.com/prometheus/prometheus/wiki/Default-port-allocations .
+	metricsAddr = ":9294"
+
+	// The HTTP path at which Prometheus metrics are served.
+	metricsPath = "/metrics"
+
+	// The namespace and subsystem of the Prometheus metrics produced here
+	metricsNamespace = "kos"
+	metricsSubsystem = "agent"
+)
+
+// layer1VirtualNetworkState is the first layer of state associated with a
+// single relevant virtual network.
+type layer1VirtualNetworkState struct {
+	// List of names of remote NetworkAttachments in the virtual network (which
+	// implicitly determines the namespace) for whom an add notification handler
+	// has been executed but a delete notification handler has not.
+	remoteAtts map[string]struct{}
+
+	// Lister used by workers to retrieve the NetworkAttachment they're
+	// processing.
+	remoteAttsLister koslisterv1a1.NetworkAttachmentNamespaceLister
+}
+
+// layer1VirtualNetworksState is the first layer of state associated with all
+// the relevant virtual networks.
+// Its main purposes are retrieval of remote NetworkAttachments by workers and
+// deletion of network interfaces of those remote NetworkAttachments when the
+// virtual network becomes irrelevant.
+// All operations on a layer1VirtualNetworksState must be done while holding
+// its mutex's lock.
+type layer1VirtualNetworksState struct {
+	sync.RWMutex
+
+	// attToListersIDs maps NetworkAttachments namespaced names to the list
+	// of IDs of the listers backed by the informers where the attachments have
+	// been seen.
+	attToListersIDs map[k8stypes.NamespacedName]map[uint32]struct{}
+
+	// vniToVNState maps a VNI to its layer1VirtualNetworkState.
+	vniToVNState map[uint32]*layer1VirtualNetworkState
+}
+
+// layer2VirtualNetworkState is the second layer of state associated with a
+// single virtual network.
+type layer2VirtualNetworkState struct {
+	// Kubernetes API namespace of the virtual network this
+	// layer2VirtualNetworkState represents.
```

Comment:
The looseness of constraint enforement in KOS means there is not necessarily just one namespace associated with a given VNI _by the collection of NetworkAttachment objects that exist_.  The discussion started at https://github.com/MikeSpreitzer/kube-examples/pull/81/files#r323568984 is part of this, but just part.  In https://github.com/MikeSpreitzer/kube-examples/tree/6a549839d890b79e7792f57cce650d66b2b51443/staging/kos#the-sdn is the accurate remark that "Neither [I think this needs updating, we now have three controllers] controller enforces or assumes any connections between the lifecycles of NetworkAttachments and their Subnets".  I reviewed the ipam controller.  It does not de-assign addresses when a Subnet is deleted.  I now think that is a bug and plan to make a PR fixing it.

Even if the ipam controller _did_ eventually de-assign addresses after a Subnet is deleted, there is a delay.  During that delay another Subnet may be created with the same VNI and a different namespace, and NetworkAttachment objects created, and even assigned addresses.

I think we can simply remove the presumption in the code that the NetworkAttachment objects associate a VNI with just one namespace.  That will, among other things, remove the possibility of an error during initial sync discussed in the discussion referenced above.

Reply:
> I reviewed the ipam controller. It does not de-assign addresses when a Subnet is deleted. I now think that is a bug and plan to make a PR fixing it.

The IPAM does not delete IPLocks when a subnet is deleted. It does, however, clear `status.IPv4` for all the attachments that were in the deleted subnet, and that means de-assigning the IP addresses **as far as the connection agent is concerned**. I agree with your following statement --- because of the delay with which IP addresses are "unassigned" it is possible for two attachments with an IP address in their status to exist and have the same VNI but different namespaces. IIUC you're proposing to stop enforcing the constraint that VNIs are confined to a single namespace, and simply let that be a possibility.  On top of my head I see nothing wrong with it, but I'd like to think this through to gain confidence that there are no catches. Anyway, I think such a battle would be worth its own PR, this PR is already unwieldy.

Reply:
I agree, let's leave that for a separate PR.